state/watcher/txnwatcher.go

```diff
@@ -0,0 +1,317 @@
+	// ChangeLog is usually the tnxs.log collection.
+	ChangeLog *mgo.Collection
+	// Hub is where the changes are published to.
+	Hub Hub
+	// Clock allows tests to control the advancing of time.
+	Clock Clock
+	// Logger is used to control where the log messages for this watcher go.
+	Logger Logger
+	// IteratorFunc can be overridden in tests to control what values the
+	// watcher sees.
+	IteratorFunc func() mongo.Iterator
+}
+
+// Validate ensures that all the values that have to be set are set.
+func (config TxnWatcherConfig) Validate() error {
+	if config.ChangeLog == nil {
+		return errors.NotValidf("missing ChangeLog")
+	}
+	if config.Hub == nil {
+		return errors.NotValidf("missing Hub")
+	}
+	if config.Clock == nil {
+		return errors.NotValidf("missing Clock")
+	}
+	return nil
+}
+
+// New returns a new Watcher observing the changelog collection,
+// which must be a capped collection maintained by mgo/txn.
+func NewTxnWatcher(config TxnWatcherConfig) (*TxnWatcher, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Annotate(err, "new TxnWatcher invalid config")
+	}
+
+	w := &TxnWatcher{
+		hub:          config.Hub,
+		clock:        config.Clock,
+		logger:       config.Logger,
+		log:          config.ChangeLog,
+		iteratorFunc: config.IteratorFunc,
+	}
+	if w.iteratorFunc == nil {
+		w.iteratorFunc = w.iter
+	}
+	if w.logger == nil {
+		w.logger = noOpLogger{}
+	}
+	go func() {
+		err := w.loop()
+		cause := errors.Cause(err)
+		// tomb expects ErrDying or ErrStillAlive as
+		// exact values, so we need to log and unwrap
+		// the error first.
+		if err != nil && cause != tomb.ErrDying {
+			w.logger.Infof("watcher loop failed: %v", err)
+		}
+		w.tomb.Kill(cause)
+		w.tomb.Done()
+	}()
+	return w, nil
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *TxnWatcher) Kill() {
+	w.tomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *TxnWatcher) Wait() error {
+	return w.tomb.Wait()
+}
+
+// Stop stops all the watcher activities.
+func (w *TxnWatcher) Stop() error {
+	return worker.Stop(w)
+}
+
+// Dead returns a channel that is closed when the watcher has stopped.
+func (w *TxnWatcher) Dead() <-chan struct{} {
+	return w.tomb.Dead()
+}
+
+// Err returns the error with which the watcher stopped.
+// It returns nil if the watcher stopped cleanly, tomb.ErrStillAlive
+// if the watcher is still running properly, or the respective error
+// if the watcher is terminating or has terminated with an error.
+func (w *TxnWatcher) Err() error {
+	return w.tomb.Err()
+}
+
+// loop implements the main watcher loop.
+// period is the delay between each sync.
+func (w *TxnWatcher) loop() error {
+	if err := w.initLastId(); err != nil {
+		return errors.Trace(err)
+	}
+	// Make sure we have read the last ID before telling people
+	// we have started.
+	w.hub.Publish(txnWatcherStarting, nil)
+	backoff := PollStrategy.NewTimer(w.clock.Now())
+	next := w.clock.After(txnWatcherShortWait)
```

Comment:
why would we always wait for the first event, shouldn't this just be After(0) ?

Reply:
In reality, the difference between waiting for 10ms and 0 is unnoticable, but for testing where you are manually advancing the clock, it gives a lot more control.